### PR TITLE
serve absolute paths to media types

### PR DESCRIPTION
**Pull request by orm011**
1) Include http server as part of runtime to serve absolute paths.
2) Provide urls to media types that resolve to this server as part of the html repr. 
3) Test src field of media types in html repr is successfully served.

I'm doing the part on user specified storage on a separate PR, as this is self-contained and fixes some things already.

- - -
This change is [Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/65)

### CODE REVIEW

File: pixeltable/dataframe.py [hunk 1]
```diff
@@ -39,26 +40,34 @@ def _format_img(img: object) -> str:
...
-    try:
-        rel_path = p.relative_to(root)
-        return f'<video controls><source src="{rel_path}" type="video/mp4"></video>'
-    except ValueError:
-        # display path as string
-        return video_file_path
+def _get_url_and_mime(file_path: str) -> Tuple[str, Optional[str]]:
+    path = Path(file_path)
+    assert path.is_absolute()
+    base = Env.get()._httpd_address.rstrip('/')
```
Comment (mkornacker) [inline pixeltable/dataframe.py]: Add an httpd_address() property to Env instead.

File: pixeltable/dataframe.py [hunk 2]
```diff
@@ -39,26 +40,34 @@ def _format_img(img: object) -> str:
...
+    url = f'{base}/{path}'
```
Comment (mkornacker) [inline pixeltable/dataframe.py]: Can you explain the url scheme somewhere?

File: pixeltable/dataframe.py [hunk 3]
```diff
@@ -39,26 +40,34 @@ def _format_img(img: object) -> str:
...
+    mime = mimetypes.guess_type(url)[0]
+    return url, mime
+
+def _format_video(file_path: str) -> str:
+    url, mime = _get_url_and_mime(file_path)
+    mime_str = f'type="{mime}"' if mime is not None else ''
```
Comment (mkornacker) [inline pixeltable/dataframe.py]: Explain in the code why this is useful.

File: pixeltable/env.py [hunk 4]
```diff
@@ -207,8 +212,27 @@ def _create_openai_client(self) -> None:
         self._logger.info('connecting to OpenAI')
         self._openai_client = openai.OpenAI(api_key=os.environ['OPENAI_API_KEY'])
 
+    def _start_web_server(self) -> None:
+        class FixedRootHandler(http.server.SimpleHTTPRequestHandler):
+            def __init__(self, *args, **kwargs):
+                super().__init__(*args, directory='/', **kwargs)
+        
+        self._httpd = socketserver.TCPServer(("127.0.0.1", 0), FixedRootHandler)
```
Comment (mkornacker) [inline pixeltable/env.py]: Port 0? Can this cause conflicts?

File: pixeltable/tests/test_dataframe.py [hunk 5]
```diff
@@ -180,6 +182,28 @@ def test_select_literal(self, test_tbl: catalog.MutableTable) -> None:
         res = t.select(1.0).where(t.c2 < 10).collect()
         assert res[res.column_names()[0]] == [1.0] * 10
 
+    def test_html_media_url(self):
```
Comment (mkornacker) [inline pixeltable/tests/test_dataframe.py]: Is there a specific reason why you're not using test_client as a parameter?

File: pixeltable/tests/test_dataframe.py [hunk 6]
```diff
@@ -180,6 +182,28 @@ def test_select_literal(self, test_tbl: catalog.MutableTable) -> None:
...
+        cl = pt.Client()
+        cl.drop_table('test_html_repr', ignore_errors=True)
+        tab = cl.create_table('test_html_repr', {'video': pt.VideoType(), 'audio': pt.AudioType()})
+        tab.insert([{'video': get_video_files()[0], 'audio': get_audio_files()[0]}])
```
Comment (mkornacker) [inline pixeltable/tests/test_dataframe.py]: status = ...
assert status.num_rows == 1
assert status.num_excs == 0

File: pixeltable/tests/test_dataframe.py [hunk 7]
```diff
@@ -180,6 +182,28 @@ def test_select_literal(self, test_tbl: catalog.MutableTable) -> None:
...
+        res = tab.select(tab.video, tab.audio).collect()
+
+        ## parse html_text to get all the video and audio tags
```
Comment (mkornacker) [inline pixeltable/tests/test_dataframe.py]: single #

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/65)* status: 0 of 6 files reviewed, 7 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/dataframe.py` line 46 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/65#gh-1456928143:-NoSARJJ8YX6CpHWsCZi:b-hk0cn1) ([raw file](https://github.com/mkornacker/pixeltable/blob/4a4230b744ba8b58d959e1e27f6424216188c02a/pixeltable/dataframe.py#L46)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Add an httpd_address() property to Env instead.
</blockquote></details>

Done. 
As a more general question: why aren't we initializing all members in the constructor?This pattern of initing things outside the constructor shows up elsewhere, but here it is especially pronounced.   I'm not sure what we gain from this pattern at least not in Python, but see a bunch of downsides when combined with other choices being made.

Because we don't init these fields,  but at the same time, we care about type annotations,  these optional fields imply constant warnings, and presumably this is one reason why we end up using properties to access simple fields, since we assert they aren't null and warnings reduce.   So we end up with a super wordy class with declarations for optional fields, then property methods for these fields,  and separate init methods.  
___
*[`pixeltable/dataframe.py` line 47 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/65#gh-1456935709:-NoSAH_BD73hZxJ8Rlim:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/4a4230b744ba8b58d959e1e27f6424216188c02a/pixeltable/dataframe.py#L47)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Can you explain the url scheme somewhere?
</blockquote></details>

Done.
___
*[`pixeltable/dataframe.py` line 53 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/65#gh-1456937779:-NoSF-_F9RAinxlLu6nw:bs7wumx) ([raw file](https://github.com/mkornacker/pixeltable/blob/4a4230b744ba8b58d959e1e27f6424216188c02a/pixeltable/dataframe.py#L53)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Explain in the code why this is useful.
</blockquote></details>

Done. 
___
*[`pixeltable/env.py` line 220 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/65#gh-1456940916:-NoSF1IKBCuKm6QhH08q:b-jin17f) ([raw file](https://github.com/mkornacker/pixeltable/blob/4a4230b744ba8b58d959e1e27f6424216188c02a/pixeltable/env.py#L220)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Port 0? Can this cause conflicts?
</blockquote></details>

lets the OS allocate an unused port for you, ie, no port conflicts. Added comment.
___
*[`pixeltable/tests/test_dataframe.py` line 185 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/65#gh-1456943596:-NoSJMiU46zFl--toAgr:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/4a4230b744ba8b58d959e1e27f6424216188c02a/pixeltable/tests/test_dataframe.py#L185)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Is there a specific reason why you're not using test_client as a parameter?
</blockquote></details>

Done.
___
*[`pixeltable/tests/test_dataframe.py` line 189 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/65#gh-1456944109:-NoSJM7I8vP6kznHQHug:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/4a4230b744ba8b58d959e1e27f6424216188c02a/pixeltable/tests/test_dataframe.py#L189)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

status = ...
assert status.num_rows == 1
assert status.num_excs == 0
</blockquote></details>

Done.
___
*[`pixeltable/tests/test_dataframe.py` line 192 at r1](https://reviewable.io/reviews/mkornacker/pixeltable/65#gh-1456946186:-NoSJNZQABsVRr-Tfqhf:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/4a4230b744ba8b58d959e1e27f6424216188c02a/pixeltable/tests/test_dataframe.py#L192)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

single #
</blockquote></details>

Done.

File: pixeltable/dataframe.py [hunk 8]
```diff
@@ -39,26 +40,33 @@ def _format_img(img: object) -> str:
...
-    # turn absolute video_file_path into relative path, absolute paths don't work
-    p = Path(video_file_path)
-    root = Path(os.getcwd())
-    try:
-        rel_path = p.relative_to(root)
-        return f'<video controls><source src="{rel_path}" type="video/mp4"></video>'
-    except ValueError:
-        # display path as string
-        return video_file_path
+def _get_src(file_path: str) -> str:
```
Comment (mkornacker) [inline pixeltable/dataframe.py]: Consider a more specific name (get_src_attr?), 'src' is very general and I had to look at the code to understand what it's returning.

File: pixeltable/dataframe.py [hunk 9]
```diff
@@ -39,26 +40,33 @@ def _format_img(img: object) -> str:
...
+    abs_path = Path(file_path)
+    assert abs_path.is_absolute()
+    # http server runs serves files wrt. to file-system root
```
Comment (mkornacker) [inline pixeltable/dataframe.py]: "runs serves"

**Review by mkornacker — APPROVED** [review]

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/65)* status: 0 of 6 files reviewed, 9 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/dataframe.py` line 46 at r2](https://reviewable.io/reviews/mkornacker/pixeltable/65#gh-1457704770:-NoSUN8VEtToUM8MGRK-:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/ffc7a762e689e2f95fd151f1e8ee348fbc4d724e/pixeltable/dataframe.py#L46)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

"runs serves"
</blockquote></details>

Done.

**Review by orm011 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/mkornacker/pixeltable/65)* status: 0 of 6 files reviewed, 9 unresolved discussions (waiting on @mkornacker)

___
*[`pixeltable/dataframe.py` line 43 at r2](https://reviewable.io/reviews/mkornacker/pixeltable/65#gh-1457703189:-NoSUOd_Cot_gNXiTz0F:b-896fix) ([raw file](https://github.com/mkornacker/pixeltable/blob/ffc7a762e689e2f95fd151f1e8ee348fbc4d724e/pixeltable/dataframe.py#L43)):*
<details><summary><i>Previously, mkornacker (Marcel Kornacker) wrote…</i></summary><blockquote>

Consider a more specific name (get_src_attr?), 'src' is very general and I had to look at the code to understand what it's returning.
</blockquote></details>

Done.

**Pull request merged**